Bot.Core/Announcer.cs

```diff
@@ -34,74 +31,45 @@ public Announcer(IServiceProvider sp)
-
-        private async void OnClientConnected(object? sender, EventArgs e)
-        {
-            await AnnounceLatestVersion();
+            m_townMaintenance.AddMaintenanceTask(AnnounceToTown);
         }
 
-        private async Task AnnounceToTowns(Queue<TownKey> towns)
+        private async Task AnnounceToTown(TownKey townKey)
         {
-            bool restricted = bool.Parse(Environment.GetEnvironmentVariable("RESTRICT_ANNOUNCE") ?? "false");
-
-            Serilog.Log.Information("AnnounceToTowns: {townCount} towns remain...", towns.Count());
-            int numSent = 0;
-            while(numSent < NUM_TOWNS_PER_CALLBACK && towns.Count > 0)
+            if(!bool.TryParse(Environment.GetEnvironmentVariable("RESTRICT_ANNOUNCE"), out bool restricted))
             {
-                var townKey = towns.Dequeue();
+                restricted = false;
+            }
 
-                // If we're restricted, skip all guilds not in the allowList
-                if (restricted && !s_guildAllowList.Contains(townKey.GuildId))
-                    continue;
+            // If we're restricted, skip all guilds not in the allowList
+            if (restricted && !s_guildAllowList.Contains(townKey.GuildId))
+                return;
 
-                Serilog.Log.Information("AnnounceToTowns: Checking town {townKey}", townKey);
+            Serilog.Log.Information("AnnounceToTown: Checking town {townKey}", townKey);
 
-                var guild = await m_botClient.GetGuildAsync(townKey.GuildId);
-                if (guild == null) 
-                    continue;
+            var guild = await m_botClient.GetGuildAsync(townKey.GuildId);
+            if (guild == null) 
+                return;
 
-                var chan = guild.GetChannel(townKey.ControlChannelId);
-                if (chan == null) 
-                    continue;
+            var chan = guild.GetChannel(townKey.ControlChannelId);
+            if (chan == null) 
+                return;
 
-                foreach (var versionObj in m_versionProvider.Versions)
+            foreach (var versionObj in m_versionProvider.Versions)
+            {
+                Serilog.Log.Information("AnnounceToTowns: Checking version {versionObj}", versionObj);
+                if (!await m_announcementDatabase.HasSeenVersion(townKey.GuildId, versionObj.Key))
                 {
-                    Serilog.Log.Information("AnnounceToTowns: Checking version {versionObj}", versionObj);
-                    if (!await m_announcementDatabase.HasSeenVersion(townKey.GuildId, versionObj.Key))
-                    {
-                        await m_announcementDatabase.RecordGuildHasSeenVersion(townKey.GuildId, versionObj.Key);
-
-                        numSent++;
-                        await chan.SendMessageAsync(versionObj.Value);
-                    }
-                }
-            }
+                    await m_announcementDatabase.RecordGuildHasSeenVersion(townKey.GuildId, versionObj.Key);
```

Comment:
This seems backward. There could be a problem sending the message (say, permissions) and here you are marking the guild as having seen the version before actually sending the message.

Reply:
Hmm. I’d rather do that then somehow keep sending a guild messages erroneously though

Reply:
Oh my assumption is that if the send fails, the message doesn't go through.

Reply:
Yeah I think you're right that your case is more likely, so I swapped these